# Import PANDAS Library

Notice that we are using pd as an abbreviation so each time we want to use the PANDAS library, we can simply type pd instead of pandas.

In [1]:
import pandas as pd

# Import Dataset (either full dataset or simplified)

You can use either set of datafiles (full or simplified). In the past some students' computers could not handle the full dataset. If you worry that your computer will not handle the full dataset (or if you find your code is running very slow), use the simplified version. 

The simplified version was created by taking a sample of the full data; hence the simplified files have "Sample" added to the end of the file names. Some files were not large to begin with and didn't need to be reduced. Those files do not have Sample added to the end of the file name.

## Adjust the file address below to match your computer.

You will want to replace the begining of the address below (i.e., everything before OneDrive - Oregon State University/)

In [2]:
begInv = pd.read_csv("BegInvFINAL12312016.csv")

In [3]:
begInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   

    Size  onHand  Price   startDate  
0  750mL       8  12.99  2016-01-01  
1  750mL       7  10.99  2016-01-01

## Deliverable 1a _______________________________________________________________________________________________________

In [4]:
begInv = begInv.rename(columns = {'onHand':'01_01_2016_Qty'})
begInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   

    Size  01_01_2016_Qty  Price   startDate  
0  750mL               8  12.99  2016-01-01  
1  750mL               7  10.99  2016-01-01

In [5]:
begInv['InvCost01_01_2016'] = begInv['01_01_2016_Qty'] * begInv['Price']
begInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   

    Size  01_01_2016_Qty  Price   startDate  InvCost01_01_2016  
0  750mL               8  12.99  2016-01-01             103.92  
1  750mL               7  10.99  2016-01-01              76.93

In [6]:
begInv_Store = begInv[['Store','01_01_2016_Qty','InvCost01_01_2016']].groupby('Store').sum()
begInv_Store.head(2)

01_01_2016_Qty  InvCost01_01_2016
Store                                   
1               49917          838079.59
2               52925          840452.20

### One way to get this is to sort the table in descending order and to use ten head(10) to see the highest. 

In [7]:
begInv_Store.sort_values(by='InvCost01_01_2016',ascending=False).head(10)

01_01_2016_Qty  InvCost01_01_2016
Store                                   
34             153852         3291170.24
73             162551         3142497.36
67             158996         3079578.63
66             149314         2973033.90
76             140208         2952418.44
69             144255         2946726.65
38             114368         2232698.77
55             119641         2001263.66
50              94720         1649808.22
79              95330         1503149.48

### Alternative way to get same answer

In [8]:
begInv_Store.sort_values(by='InvCost01_01_2016').tail(10).sort_values(by='InvCost01_01_2016', ascending = False)

01_01_2016_Qty  InvCost01_01_2016
Store                                   
34             153852         3291170.24
73             162551         3142497.36
67             158996         3079578.63
66             149314         2973033.90
76             140208         2952418.44
69             144255         2946726.65
38             114368         2232698.77
55             119641         2001263.66
50              94720         1649808.22
79              95330         1503149.48

## Deliverable 1b _______________________________________________________________________________________________________

In [9]:
begInv_Brand = begInv[['Brand','01_01_2016_Qty','InvCost01_01_2016']].groupby('Brand').sum()
begInv_Brand.head(2)

01_01_2016_Qty  InvCost01_01_2016
Brand                                   
58                281            3650.19
60                288            3165.12

In [10]:
begInv_Brand.sort_values(by='InvCost01_01_2016',ascending=False).head(10)

01_01_2016_Qty  InvCost01_01_2016
Brand                                   
3545            14499          463823.01
1233            12016          432455.84
8068            15341          383371.59
4261            15499          340823.01
3858            13649          327439.51
2753             4625          286703.75
8082             9287          278517.13
8680             7066          275503.34
2589             6766          270572.34
3876            14829          266773.71

### Alternative way to get same answer

In [11]:
begInv_Brand.sort_values(by='InvCost01_01_2016').tail(10).sort_values(by='InvCost01_01_2016', ascending = False)

01_01_2016_Qty  InvCost01_01_2016
Brand                                   
3545            14499          463823.01
1233            12016          432455.84
8068            15341          383371.59
4261            15499          340823.01
3858            13649          327439.51
2753             4625          286703.75
8082             9287          278517.13
8680             7066          275503.34
2589             6766          270572.34
3876            14829          266773.71

## Deliverable 2a _______________________________________________________________________________________________________

In [12]:
endInv = pd.read_csv("EndInvFINAL12312016.csv")
endInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   

    Size  onHand  Price     endDate  
0  750mL      11  12.99  2016-12-31  
1  750mL       7  36.99  2016-12-31

In [13]:
endInv = endInv.rename(columns = {'onHand':'12_31_2016_Qty'})
endInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   

    Size  12_31_2016_Qty  Price     endDate  
0  750mL              11  12.99  2016-12-31  
1  750mL               7  36.99  2016-12-31

In [14]:
endInv['InvCost12_31_2016'] = endInv['12_31_2016_Qty'] * endInv['Price']
endInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   

    Size  12_31_2016_Qty  Price     endDate  InvCost12_31_2016  
0  750mL              11  12.99  2016-12-31             142.89  
1  750mL               7  36.99  2016-12-31             258.93

In [15]:
endInv_Store = endInv[['Store','12_31_2016_Qty','InvCost12_31_2016']].groupby('Store').sum()
endInv_Store.head(2)

12_31_2016_Qty  InvCost12_31_2016
Store                                   
1               79827         1206845.93
2               56671          850884.06

### One way to get this is to sort the table in descending order and to use ten head(10) to see the highest. 

In [16]:
endInv_Store.sort_values(by='InvCost12_31_2016',ascending=False).head(10)

12_31_2016_Qty  InvCost12_31_2016
Store                                   
50             260717         4887260.68
73             164589         3254662.81
67             163765         3076114.82
34             145829         3074616.75
76             143866         2975945.18
69             150848         2968678.82
66             144579         2860504.99
74             166015         2803645.13
38             129397         2463906.85
55             125584         2234836.35

### Alternative way to get same answer

In [17]:
endInv_Store.sort_values(by='InvCost12_31_2016').tail(10).sort_values(by='InvCost12_31_2016', ascending = False)

12_31_2016_Qty  InvCost12_31_2016
Store                                   
50             260717         4887260.68
73             164589         3254662.81
67             163765         3076114.82
34             145829         3074616.75
76             143866         2975945.18
69             150848         2968678.82
66             144579         2860504.99
74             166015         2803645.13
38             129397         2463906.85
55             125584         2234836.35

## Deliverable 2b _______________________________________________________________________________________________________

In [18]:
endInv_Brand = endInv[['Brand','12_31_2016_Qty','InvCost12_31_2016']].groupby('Brand').sum()
endInv_Brand.head(2)

12_31_2016_Qty  InvCost12_31_2016
Brand                                   
58                385            5001.15
60                146            1604.54

In [19]:
endInv_Brand.sort_values(by='InvCost12_31_2016',ascending=False).head(10)

12_31_2016_Qty  InvCost12_31_2016
Brand                                   
1233            15047          526494.53
3545            16770          502932.30
2753             7849          470861.51
8068            15608          366631.92
3405            12268          355649.32
4261            16769          351981.31
2757            11603          336370.97
2589             7922          300956.78
1376            13180          276648.20
2585            10487          272557.13

### Alternative way to get same answer

In [20]:
endInv_Brand.sort_values(by='InvCost12_31_2016').tail(10).sort_values(by='InvCost12_31_2016', ascending = False)

12_31_2016_Qty  InvCost12_31_2016
Brand                                   
1233            15047          526494.53
3545            16770          502932.30
2753             7849          470861.51
8068            15608          366631.92
3405            12268          355649.32
4261            16769          351981.31
2757            11603          336370.97
2589             7922          300956.78
1376            13180          276648.20
2585            10487          272557.13

## Deliverable 3
_______________________________________________________________________________________________________

In [21]:
threeBegInv = pd.read_csv("BegInvFINAL12312016.csv")

threeBegInv = threeBegInv.rename(columns = {'onHand':'BB#ofUnits'})

threeBegInv['BB$Cost'] = threeBegInv['BB#ofUnits'] * threeBegInv['Price']

threeBegInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_60      1  HARDERSFIELD     60       Canadian Club 1858 VAP   

    Size  BB#ofUnits  Price   startDate  BB$Cost  
0  750mL           8  12.99  2016-01-01   103.92  
1  750mL           7  10.99  2016-01-01    76.93

In [22]:
threeEndInv = pd.read_csv("EndInvFINAL12312016.csv")

threeEndInv = threeEndInv.rename(columns = {'onHand':'EB#ofUnits'})

threeEndInv['EB$Cost'] = threeEndInv['EB#ofUnits'] * threeEndInv['Price']

threeEndInv.head(2)

InventoryId  Store          City  Brand                  Description  \
0  1_HARDERSFIELD_58      1  HARDERSFIELD     58  Gekkeikan Black & Gold Sake   
1  1_HARDERSFIELD_62      1  HARDERSFIELD     62     Herradura Silver Tequila   

    Size  EB#ofUnits  Price     endDate  EB$Cost  
0  750mL          11  12.99  2016-12-31   142.89  
1  750mL           7  36.99  2016-12-31   258.93

In [23]:
mergedInv = pd.merge(left=threeBegInv, right=threeEndInv, how='outer', on=['InventoryId']).fillna(0)

simplifiedmergedInv = mergedInv[["InventoryId", "BB#ofUnits", "BB$Cost", "EB#ofUnits", "EB$Cost"]]

simplifiedmergedInv.head()

InventoryId  BB#ofUnits  BB$Cost  EB#ofUnits  EB$Cost
0  1_HARDERSFIELD_58         8.0   103.92        11.0   142.89
1  1_HARDERSFIELD_60         7.0    76.93         0.0     0.00
2  1_HARDERSFIELD_62         6.0   221.94         7.0   258.93
3  1_HARDERSFIELD_63         3.0   116.97         7.0   272.93
4  1_HARDERSFIELD_72         6.0   209.94         4.0   139.96

In [24]:
len(simplifiedmergedInv)

256042

## Deliverable 4a
_______________________________________________________________________________________________________

In [25]:
fourPurchases = pd.read_csv("PurchasesFINAL12312016.csv")

fourPurchases.head(2)

InventoryId  Store  Brand                   Description   Size  \
0  69_MOUNTMEND_8412     69   8412     Tequila Ocho Plata Fresno  750mL   
1   30_CULCHETH_5255     30   5255  TGI Fridays Ultimte Mudslide  1.75L   

   VendorNumber                   VendorName  PONumber      PODate  \
0           105  ALTAMAR BRANDS LLC               8124  2015-12-21   
1          4466  AMERICAN VINTAGE BEVERAGE        8137  2015-12-22   

  ReceivingDate InvoiceDate     PayDate  PurchasePrice  Quantity  Dollars  \
0    2016-01-02  2016-01-04  2016-02-16          35.71         6   214.26   
1    2016-01-01  2016-01-07  2016-02-21           9.35         4    37.40   

   Classification  
0               1  
1               1

In [26]:
fourPurchases[["VendorNumber", "Dollars"]].groupby("VendorNumber").sum().sort_values(by="Dollars", ascending=False).head(10)

Dollars
VendorNumber             
3960          50959796.85
4425          27861690.02
12546         24203151.05
17035         24124091.56
480           17624378.72
1392          15573917.90
1128          13529433.08
9165          13210613.93
3252          12289608.09
9552          10935817.30

## Deliverable 4b
_______________________________________________________________________________________________________

In [27]:
fourInvoices = pd.read_csv("InvoicePurchases12312016.csv")

fourInvoices[["VendorNumber", "Freight"]].groupby("VendorNumber").sum().sort_values(by="Freight", ascending=False).head(10)

Freight
VendorNumber           
3960          257032.07
4425          144929.24
12546         123880.97
17035         123780.22
480            89286.27
1392           79528.99
1128           68601.68
9165           68054.70
3252           61966.91
9552           55551.82

## Deliverable 4c
_______________________________________________________________________________________________________

In [28]:
LargeVend = fourInvoices.groupby(["VendorNumber","VendorName"])[["Freight","Dollars","Quantity"]].sum()

LargeVend['Per$Freight'] = LargeVend['Freight'] / LargeVend['Dollars']

LargeVend = LargeVend[LargeVend["Dollars"]>250000]

LargeVend.head()
                                                                

Freight      Dollars  Quantity  \
VendorNumber VendorName                                                     
480          BACARDI USA INC              89286.27  17624378.72   1427075   
516          BANFI PRODUCTS CORP           8510.41   1628866.68    228103   
653          STATE WINE & SPIRITS          8014.98   1529682.04    154092   
660          SAZERAC NORTH AMERICA INC.   17932.33   3537977.55    503931   
1128         BROWN-FORMAN CORP            68601.68  13529433.08   1006122   

                                          Per$Freight  
VendorNumber VendorName                                
480          BACARDI USA INC                 0.005066  
516          BANFI PRODUCTS CORP             0.005225  
653          STATE WINE & SPIRITS            0.005240  
660          SAZERAC NORTH AMERICA INC.      0.005069  
1128         BROWN-FORMAN CORP               0.005071

In [29]:
LargeVend.sort_values(by="Per$Freight", ascending=False).head(5) #for the 5 highest per $ freight

Freight     Dollars  Quantity  \
VendorNumber VendorName                                                   
9625         WESTERN SPIRITS BEVERAGE CO  1933.19   361249.21     56860   
1590         DIAGEO CHATEAU ESTATE WINES  7259.75  1365472.83    187841   
9744         FREDERICK WILDMAN & SONS     3999.93   759449.24     70932   
653          STATE WINE & SPIRITS         8014.98  1529682.04    154092   
17031        FLAG HILL WINERY & VINEYARD  1573.31   300403.20     20608   

                                          Per$Freight  
VendorNumber VendorName                                
9625         WESTERN SPIRITS BEVERAGE CO     0.005351  
1590         DIAGEO CHATEAU ESTATE WINES     0.005317  
9744         FREDERICK WILDMAN & SONS        0.005267  
653          STATE WINE & SPIRITS            0.005240  
17031        FLAG HILL WINERY & VINEYARD     0.005237

In [30]:
LargeVend.sort_values(by="Per$Freight", ascending=True).head(5) #for the 5 lowest per $ freight

Freight     Dollars  Quantity  \
VendorNumber VendorName                                                    
6359         OLE SMOKY DISTILLERY LLC      1922.00   387622.69    110162   
9819         TREASURY WINE ESTATES        14836.57  2978686.40    497770   
3089         SIDNEY FRANK IMPORTING CO     8549.55  1715908.88    186464   
3924         HEAVEN HILL DISTILLERIES     14069.87  2816661.94    352348   
8352         LUXCO INC                    10261.60  2051436.01    283260   

                                          Per$Freight  
VendorNumber VendorName                                
6359         OLE SMOKY DISTILLERY LLC        0.004958  
9819         TREASURY WINE ESTATES           0.004981  
3089         SIDNEY FRANK IMPORTING CO       0.004983  
3924         HEAVEN HILL DISTILLERIES        0.004995  
8352         LUXCO INC                       0.005002

## Deliverable 4d
_______________________________________________________________________________________________________

In [31]:
filterunitfreight = fourInvoices[(fourInvoices["Freight"] > 100) & (fourInvoices["Quantity"] <= 1000)]

filterunitfreight.sort_values(by="Freight", ascending=False).head(5) #to show the top 5 freight cost under these constraints

VendorNumber                   VendorName InvoiceDate  PONumber  \
71           8664  STOLI GROUP,(USA) LLC        2016-01-12      8202   
59           7239  REMY COINTREAU USA INC       2016-01-08      8163   
69            653      STATE WINE & SPIRITS     2016-01-08      8201   
50           6785  PALM BAY INTERNATIONAL INC   2016-01-08      8114   
407          2561  EDRINGTON AMERICAS           2016-02-01      8591   

         PODate     PayDate  Quantity   Dollars  Freight Approval  
71   2015-12-25  2016-02-04       975  14299.98   349.02      NaN  
59   2015-12-23  2016-02-18       382   7515.38   348.07      NaN  
69   2015-12-25  2016-02-13       541   5544.89   303.70      NaN  
50   2015-12-20  2016-02-14       715   5533.18   242.73      NaN  
407  2016-01-22  2016-03-17       765  42239.46   219.65      NaN

As seen above, the vendors with the highest freight costs under these conditions included vendor 8664, 7239, 653, 6785, and 2561.

## Deliverable 4e
_______________________________________________________________________________________________________

In [32]:
SmallExpensive = fourInvoices[(fourInvoices["Freight"] > 100) & (fourInvoices["Quantity"] <= 1000)]

SmallExpensive['Per$Freight'] = SmallExpensive['Freight'] / SmallExpensive['Dollars']

SmallExpensive['PerUnitFreight'] = SmallExpensive['Freight'] / SmallExpensive['Quantity']

SmallExpensive.head()


C:\Users\19716\AppData\Local\Temp\ipykernel_10612\2607608658.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  SmallExpensive['Per$Freight'] = SmallExpensive['Freight'] / SmallExpensive['Dollars']
C:\Users\19716\AppData\Local\Temp\ipykernel_10612\2607608658.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  SmallExpensive['PerUnitFreight'] = SmallExpensive['Freight'] / SmallExpensive['Quantity']


VendorNumber                   VendorName InvoiceDate  PONumber  \
11          1485  CASTLE BRANDS CORP.          2016-01-08      8152   
16          2242  DELICATO VINEYARDS INC       2016-01-06      8139   
19          2555  DISARONNO INTERNATIONAL LLC  2016-01-11      8192   
23          2561  EDRINGTON AMERICAS           2016-01-08      8175   
28          3924  HEAVEN HILL DISTILLERIES     2016-01-08      8155   

        PODate     PayDate  Quantity  Dollars  Freight Approval  Per$Freight  \
11  2015-12-23  2016-02-19       320  5420.41   179.26      NaN     0.033071   
16  2015-12-22  2016-02-10       808  6646.46   127.05      NaN     0.019115   
19  2015-12-25  2016-02-17       385  3506.41   146.86      NaN     0.041883   
23  2015-12-24  2016-02-10       136  5645.24   218.18      NaN     0.038648   
28  2015-12-23  2016-02-15       818  7079.02   200.02      NaN     0.028255   

    PerUnitFreight  
11        0.560187  
16        0.157240  
19        0.381455  
23        1.604265  
28        0.244523

In [33]:
LargeVend = fourInvoices.groupby(["VendorNumber","VendorName"])[["Freight","Dollars","Quantity"]].sum()

LargeVend = LargeVend[LargeVend["Dollars"]>250000]

LargeVend['Per$Freight'] = LargeVend['Freight'] / LargeVend['Dollars']

LargeVend['PerUnitFreight'] = LargeVend['Freight'] / LargeVend['Quantity']

LargeVend = LargeVend[LargeVend["Dollars"]>250000]

LargeVend.head()

Freight      Dollars  Quantity  \
VendorNumber VendorName                                                     
480          BACARDI USA INC              89286.27  17624378.72   1427075   
516          BANFI PRODUCTS CORP           8510.41   1628866.68    228103   
653          STATE WINE & SPIRITS          8014.98   1529682.04    154092   
660          SAZERAC NORTH AMERICA INC.   17932.33   3537977.55    503931   
1128         BROWN-FORMAN CORP            68601.68  13529433.08   1006122   

                                          Per$Freight  PerUnitFreight  
VendorNumber VendorName                                                
480          BACARDI USA INC                 0.005066        0.062566  
516          BANFI PRODUCTS CORP             0.005225        0.037310  
653          STATE WINE & SPIRITS            0.005240        0.052014  
660          SAZERAC NORTH AMERICA INC.      0.005069        0.035585  
1128         BROWN-FORMAN CORP               0.005071        0.068184

In [34]:
avg_small_expensive = SmallExpensive[['Per$Freight', 'PerUnitFreight']].mean()
avg_large_vend = LargeVend[['Per$Freight', 'PerUnitFreight']].mean()

MeanTable = pd.DataFrame([avg_small_expensive, avg_large_vend], index=['Small and Expensive', 'Large Vendors'])

MeanTableTransp = MeanTable.T

MeanTableTransp

Small and Expensive  Large Vendors
Per$Freight                0.016157        0.00510
PerUnitFreight             0.274671        0.05439

Comparing these two, I see that the small/expensive vendors have a much higher freight rate (for quantity and dollars) relative to the large vendors, who have a relatively smaller freight rate (for quantity and dollars). This makes sense, as larger orders might get better freight prices/bulk discount shipping deals. 

## Deliverable 5: Freeform Sales and Purchases data combination and analysis. 
_______________________________________________________________________________________________________

My goal is to show the difference by brand in amounts recorded as received in Sales data versus dollars invoiced from the Invoice Purchases data. I am making the simplified assumption here that cash inflow from sales would be the former, and the difference between the two would be accounts receivable. This might be helpful to show the brands most purchased on credit by dollar amount.

In [35]:
Sales = pd.read_csv("SalesFINAL12312016.csv")

ByVenSales = Sales.groupby(["VendorNo","VendorName"])[["SalesDollars"]].sum().sort_values(by="SalesDollars", ascending=False)

ByVenSales.rename(columns={'SalesDollars': 'SalesDollarsRecorded'}, inplace=True)

ByVenSales.head()


SalesDollarsRecorded
VendorNo VendorName                                       
3960     DIAGEO NORTH AMERICA INC              68742416.99
4425     MARTIGNETTI COMPANIES                 40992395.93
17035    PERNOD RICARD USA                     32281247.95
12546    JIM BEAM BRANDS COMPANY               31906320.54
480      BACARDI USA INC                       25014556.89

In [36]:
InvPurch = pd.read_csv("InvoicePurchases12312016.csv")

ByVenInvPurch = InvPurch.groupby(["VendorNumber","VendorName"])[["Dollars"]].sum().sort_values(by="Dollars", ascending=False)

ByVenInvPurch.rename(columns={'Dollars': 'DollarsInvoiced'}, inplace=True)

ByVenInvPurch.head()


DollarsInvoiced
VendorNumber VendorName                                  
3960         DIAGEO NORTH AMERICA INC         50959796.85
4425         MARTIGNETTI COMPANIES            27821473.91
12546        JIM BEAM BRANDS COMPANY          24203151.05
17035        PERNOD RICARD USA                24124091.56
480          BACARDI USA INC                  17624378.72

In [37]:
mergedVens = pd.merge(left=ByVenInvPurch, right=ByVenSales, how='outer', on=['VendorName'])

mergedVens['InvoicedOverSale'] = mergedVens['DollarsInvoiced'] - mergedVens['SalesDollarsRecorded']

mergedVens.sort_values(by="InvoicedOverSale", ascending=False).head()




DollarsInvoiced  \
VendorName                                                 
ADAMBA IMPORTS INTL INC                         76770.25   
BLACK COVE BEVERAGES                            14465.06   
VINEYARD BRANDS LLC                             10951.51   
IRA GOLDMAN AND WILLIAMS, LLP                    5630.88   
HIGHLAND WINE MERCHANTS LLC                      5500.32   

                                         SalesDollarsRecorded  \
VendorName                                                      
ADAMBA IMPORTS INTL INC                              67576.22   
BLACK COVE BEVERAGES                                  6256.87   
VINEYARD BRANDS LLC                                   4323.96   
IRA GOLDMAN AND WILLIAMS, LLP                         1265.58   
HIGHLAND WINE MERCHANTS LLC                           1533.68   

                                         InvoicedOverSale  
VendorName                                                 
ADAMBA IMPORTS INTL INC                           9194.03  
BLACK COVE BEVERAGES                              8208.19  
VINEYARD BRANDS LLC                               6627.55  
IRA GOLDMAN AND WILLIAMS, LLP                     4365.30  
HIGHLAND WINE MERCHANTS LLC                       3966.64

The above 5 brands represent the top 5 that A/R balances can be attributed to.